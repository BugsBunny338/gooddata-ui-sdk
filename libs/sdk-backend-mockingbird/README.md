# Mocking bird

This package contains test support and mock implementations of various concepts in the sdk-backend-spi.
