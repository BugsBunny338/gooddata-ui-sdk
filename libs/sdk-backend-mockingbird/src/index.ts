// (C) 2019-2020 GoodData Corporation

import {
    dummyBackend,
    dummyBackendEmptyData,
    dummyDataView,
    dummyDataFacade,
} from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView, dummyDataFacade };

export { recordedBackend } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
} from "./recordedBackend/types";

export {
    recordedDataView,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
} from "./recordedBackend/execution";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataFacade,
} from "./legacyRecordedBackend";
