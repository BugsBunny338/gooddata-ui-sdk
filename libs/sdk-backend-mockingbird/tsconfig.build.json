{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "baseUrl": null,
        "paths": null
    },
    "include": ["src/index.ts"]
}
