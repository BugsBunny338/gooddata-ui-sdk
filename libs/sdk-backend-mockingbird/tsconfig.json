{
    "compilerOptions": {
        "esModuleInterop": true,
        "target": "es5",
        "module": "commonjs",
        "lib": ["es6", "dom"],
        "declaration": true,
        "outDir": "./dist",
        "strict": true,
        "noImplicitAny": true,
        "strictNullChecks": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "baseUrl": "..",
        "paths": {
            "@gooddata/sdk-model": ["sdk-model/src"],
            "@gooddata/sdk-backend-base": ["sdk-backend-base/src"],
            "@gooddata/sdk-backend-spi": ["sdk-backend-spi/src"]
        }
    },
    "exclude": ["node_modules", "dist"]
}
