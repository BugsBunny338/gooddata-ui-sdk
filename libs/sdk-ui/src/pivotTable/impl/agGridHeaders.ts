// (C) 2007-2019 GoodData Corporation

import { getMappingHeaderName } from "../../base/helpers/mappingHeader";
import { unwrap } from "../../base/helpers/utils";
import { IMappingHeader } from "../../base/interfaces/MappingHeader";
import { getIdsFromUri, getTreeLeaves } from "./agGridUtils";
import {
    COLUMN_GROUPING_DELIMITER,
    FIELD_SEPARATOR,
    ID_SEPARATOR,
    ROW_ATTRIBUTE_COLUMN,
} from "./agGridConst";
import { assignDrillItemsAndType } from "./agGridDrilling";
import { IColumnDefOptions, IGridAdapterOptions, IGridHeader, TableHeaders } from "./agGridTypes";
import {
    DataValue,
    DataViewFacade,
    IAttributeDescriptor,
    IDataView,
    IDimensionItemDescriptor,
    IMeasureDescriptor,
    IDimensionDescriptor,
    IResultHeader,
    isAttributeDescriptor,
    isMeasureGroupDescriptor,
    isResultAttributeHeader,
    isResultMeasureHeader,
    isResultTotalHeader,
} from "@gooddata/sdk-backend-spi";
import {
    IAttributeSortItem,
    IMeasureSortItem,
    isAttributeSort,
    isMeasureLocator,
    isMeasureSort,
} from "@gooddata/sdk-model";
import { ColDef } from "ag-grid-community";
import range = require("lodash/range");
import clone = require("lodash/clone");
import invariant = require("invariant");

/*
 * All code related to transforming headers from our backend to ag-grid specific data structures
 */

export const identifyHeader = (header: IResultHeader) => {
    if (isResultAttributeHeader(header)) {
        return `a${ID_SEPARATOR}${getIdsFromUri(header.attributeHeaderItem.uri).join(ID_SEPARATOR)}`;
    }
    if (isResultMeasureHeader(header)) {
        return `m${ID_SEPARATOR}${header.measureHeaderItem.order}`;
    }
    if (isResultTotalHeader(header)) {
        return `t${ID_SEPARATOR}${header.totalHeaderItem.type}`;
    }
    invariant(false, `Unknown header type: ${JSON.stringify(header)}`);
};

export const identifyResponseHeader = (header: IDimensionItemDescriptor) => {
    if (isAttributeDescriptor(header)) {
        // response headers have no value id
        return `a${ID_SEPARATOR}${getIdsFromUri(header.attributeHeader.uri)[0]}`;
    }
    if (isMeasureGroupDescriptor(header)) {
        // trying to identify a measure group would be ambiguous
        return null;
    }
    invariant(false, `Unknown response header type: ${JSON.stringify(header)}`);
};

export const headerToGrid = (header: IResultHeader, fieldPrefix = "") => {
    const internalHeader = unwrap(header);
    return {
        headerName: internalHeader.name,
        field: fieldPrefix + identifyHeader(header),
    };
};

export const shouldMergeHeaders = (
    resultHeaderDimension: IResultHeader[][],
    headerIndex: number,
    headerItemIndex: number,
): boolean => {
    for (let ancestorIndex = headerIndex; ancestorIndex >= 0; ancestorIndex--) {
        const currentAncestorHeader = resultHeaderDimension[ancestorIndex][headerItemIndex];
        const nextAncestorHeader = resultHeaderDimension[ancestorIndex][headerItemIndex + 1];
        if (
            !nextAncestorHeader ||
            identifyHeader(currentAncestorHeader) !== identifyHeader(nextAncestorHeader)
        ) {
            return false;
        }
    }
    return true;
};

export const mergeHeaderEndIndex = (
    resultHeaderDimension: IResultHeader[][],
    headerIndex: number,
    headerItemStartIndex: number,
): number => {
    const header = resultHeaderDimension[headerIndex];
    for (let headerItemIndex = headerItemStartIndex; headerItemIndex < header.length; headerItemIndex++) {
        if (!shouldMergeHeaders(resultHeaderDimension, headerIndex, headerItemIndex)) {
            return headerItemIndex;
        }
    }
    return headerItemStartIndex;
};

/*
 * getColumnHeaders transforms header items from matrix to tree hierarchy
 *Â for each span of identical headers in a row, the function is called recursively to assign child items
 */
export const getColumnHeaders = (
    resultHeaderDimension: IResultHeader[][],
    responseHeaders: IDimensionItemDescriptor[],
    columnDefOptions: IColumnDefOptions = {},
    headerIndex = 0,
    headerItemStartIndex = 0,
    headerValueEnd: number = undefined,
    fieldPrefix = "",
    parentDrillItems: IMappingHeader[] = [],
): IGridHeader[] => {
    if (!resultHeaderDimension.length) {
        return [];
    }

    const currentHeaders = resultHeaderDimension[headerIndex];
    const lastIndex = headerValueEnd !== undefined ? headerValueEnd : currentHeaders.length - 1;
    const hierarchy: IGridHeader[] = [];

    for (let headerItemIndex = headerItemStartIndex; headerItemIndex < lastIndex + 1; ) {
        const currentHeader = currentHeaders[headerItemIndex];
        const header: IGridHeader = {
            drillItems: [],
            ...headerToGrid(currentHeader, fieldPrefix),
            ...columnDefOptions,
        };
        const drillItems: IMappingHeader[] = clone(parentDrillItems);
        assignDrillItemsAndType(header, currentHeader, responseHeaders, headerIndex, drillItems);
        const headerItemEndIndex = mergeHeaderEndIndex(resultHeaderDimension, headerIndex, headerItemIndex);

        if (headerIndex !== resultHeaderDimension.length - 1) {
            header.children = getColumnHeaders(
                resultHeaderDimension,
                responseHeaders,
                columnDefOptions,
                headerIndex + 1,
                headerItemIndex,
                headerItemEndIndex,
                header.field + FIELD_SEPARATOR,
                drillItems,
            );
        }
        hierarchy.push(header);
        // We move the pointer manually to skip identical headers
        headerItemIndex = headerItemEndIndex + 1;
    }

    return hierarchy;
};

export const getRowHeaders = (
    rowDimensionHeaders: IAttributeDescriptor[],
    columnDefOptions: IColumnDefOptions,
    makeRowGroups: boolean,
): IGridHeader[] => {
    return rowDimensionHeaders.map((attributeHeader: IAttributeDescriptor) => {
        const rowGroupProps = makeRowGroups
            ? {
                  rowGroup: true,
                  hide: true,
              }
            : {};
        const field = identifyResponseHeader(attributeHeader);
        return {
            // The label should be attribute name (not attribute display form name)
            headerName: getMappingHeaderName(attributeHeader),
            type: ROW_ATTRIBUTE_COLUMN,
            // Row dimension must contain only attribute headers.
            field,
            drillItems: [attributeHeader],
            ...rowGroupProps,
            ...columnDefOptions,
        };
    });
};

export const getFields = (dataHeaders: IResultHeader[][]) => {
    return range((dataHeaders[0] || []).length).map((cellIndex: number) => {
        const fieldList = dataHeaders.map((header: IResultHeader[]) => identifyHeader(header[cellIndex]));
        return fieldList.join(FIELD_SEPARATOR);
    }) as string[];
};

// TODO: move this to data view facade / sanitize / make more generic
export const assortDimensionHeaders = (dimensions: IDimensionDescriptor[]) => {
    const dimensionHeaders: IDimensionItemDescriptor[] = dimensions.reduce(
        (headers: IDimensionItemDescriptor[], dimension: IDimensionDescriptor) => [
            ...headers,
            ...dimension.headers,
        ],
        [],
    );
    const attributeHeaders: IAttributeDescriptor[] = [];
    const measureHeaderItems: IMeasureDescriptor[] = [];
    dimensionHeaders.forEach((dimensionHeader: IDimensionItemDescriptor) => {
        if (isAttributeDescriptor(dimensionHeader)) {
            attributeHeaders.push(dimensionHeader);
        } else if (isMeasureGroupDescriptor(dimensionHeader)) {
            measureHeaderItems.push(...dimensionHeader.measureGroupHeader.items);
        }
    });
    return {
        attributeHeaders,
        measureHeaderItems,
    };
};

export const getMinimalRowData = (data: DataValue[][], rowHeaderItems: IResultHeader[][]) => {
    const numberOfRowHeaderItems = (rowHeaderItems[0] || []).length;

    return data.length > 0
        ? data
        : // if there are no measures only attributes
          // create array of [null] of length equal to the number of row dimension headerItems
          (Array(numberOfRowHeaderItems).fill([null]) as DataValue[][]);
};

const assignSorting = (colDef: ColDef, sortingMap: { [key: string]: string }): void => {
    const direction = sortingMap[colDef.field];
    if (direction) {
        colDef.sort = direction;
    }
};

/**
 * exported due to tests
 * @internal
 */
export const getAttributeSortItemFieldAndDirection = (
    sortItem: IAttributeSortItem,
    attributeHeaders: IAttributeDescriptor[],
): [string, string] => {
    const localIdentifier = sortItem.attributeSortItem.attributeIdentifier;

    const sortHeader = attributeHeaders.find(
        header => header.attributeHeader.localIdentifier === localIdentifier,
    );
    invariant(sortHeader, `Could not find sortHeader with localIdentifier ${localIdentifier}`);

    const field = identifyResponseHeader(sortHeader);
    return [field, sortItem.attributeSortItem.direction];
};

/**
 * exported due to tests
 * @internal
 */
export const getMeasureSortItemFieldAndDirection = (
    sortItem: IMeasureSortItem,
    measureHeaderItems: IMeasureDescriptor[],
): [string, string] => {
    const keys: string[] = [];
    sortItem.measureSortItem.locators.forEach(locator => {
        if (isMeasureLocator(locator)) {
            const measureSortHeaderIndex = measureHeaderItems.findIndex(
                measureHeaderItem =>
                    measureHeaderItem.measureHeaderItem.localIdentifier ===
                    locator.measureLocatorItem.measureIdentifier,
            );
            keys.push(`m${ID_SEPARATOR}${measureSortHeaderIndex}`);
        } else {
            const key = `a${ID_SEPARATOR}${getIdsFromUri(locator.attributeLocatorItem.element).join(
                ID_SEPARATOR,
            )}`;
            keys.push(key);
        }
    });
    const field = keys.join(FIELD_SEPARATOR);
    const direction = sortItem.measureSortItem.direction;
    return [field, direction];
};

export function createTableHeaders(dataView: IDataView, options: IGridAdapterOptions = {}): TableHeaders {
    const dv = new DataViewFacade(dataView);
    const dimensions = dv.dimensions();
    const headerItems = dv.headerItems();
    const { columnDefOptions, makeRowGroups = false } = options;

    const sorting = dv.definition.sortBy;
    const sortingMap = {};
    const { attributeHeaders, measureHeaderItems } = assortDimensionHeaders(dimensions);
    sorting.forEach(sortItem => {
        if (isAttributeSort(sortItem)) {
            const [field, direction] = getAttributeSortItemFieldAndDirection(sortItem, attributeHeaders);
            sortingMap[field] = direction;
        }
        if (isMeasureSort(sortItem)) {
            const [field, direction] = getMeasureSortItemFieldAndDirection(sortItem, measureHeaderItems);
            sortingMap[field] = direction;
        }
    });

    const columnAttributeHeaderCount = dimensions[1].headers.filter(
        (header: IDimensionItemDescriptor) => !!(header as IAttributeDescriptor).attributeHeader,
    ).length;

    const columnHeaders: IGridHeader[] = getColumnHeaders(
        headerItems[1],
        dimensions[1].headers,
        columnDefOptions,
    );

    const groupColumnHeaders: IGridHeader[] =
        columnAttributeHeaderCount > 0
            ? [
                  {
                      headerName: dimensions[1].headers
                          .filter(header => isAttributeDescriptor(header))
                          .map((header: IAttributeDescriptor) => {
                              return getMappingHeaderName(header);
                          })
                          .filter((item: string) => item !== null)
                          .join(COLUMN_GROUPING_DELIMITER),
                      field: "columnGroupLabel",
                      children: columnHeaders,
                      drillItems: [],
                  },
              ]
            : columnHeaders;

    const rowHeaders: IGridHeader[] =
        // There are supposed to be only attribute headers on the first dimension
        getRowHeaders(dimensions[0].headers as IAttributeDescriptor[], columnDefOptions, makeRowGroups);

    const allHeaders: IGridHeader[] = [...rowHeaders, ...groupColumnHeaders].map((column, index) => {
        if (column.children) {
            getTreeLeaves(column).forEach((leafColumn, leafColumnIndex) => {
                leafColumn.index = index + leafColumnIndex;
                assignSorting(leafColumn, sortingMap);
            });
        }
        column.index = index;
        assignSorting(column, sortingMap);
        return column;
    });

    const colFields: string[] = getFields(headerItems[1]);
    const rowFields: string[] = rowHeaders.map(header => header.field);

    const leafColumnDefs = getTreeLeaves(allHeaders);
    if (leafColumnDefs[0]) {
        leafColumnDefs[0].cellRenderer = "loadingRenderer";
    }

    return {
        rowHeaders,
        colHeaders: groupColumnHeaders,
        allHeaders,
        rowFields,
        colFields,
    };
}
