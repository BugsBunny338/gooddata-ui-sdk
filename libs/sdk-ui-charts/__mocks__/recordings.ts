// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-backend-spi";
import { ScenarioRecording, recordedDataView } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording): DataViewFacade {
    return new DataViewFacade(recordedDataView(rec));
}
