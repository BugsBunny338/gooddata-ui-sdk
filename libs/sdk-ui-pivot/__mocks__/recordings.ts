// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-backend-spi";
import { ScenarioRecording, recordedDataView, DataViewAll } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording, dataViewId: string = DataViewAll): DataViewFacade {
    return new DataViewFacade(recordedDataView(rec, dataViewId));
}
