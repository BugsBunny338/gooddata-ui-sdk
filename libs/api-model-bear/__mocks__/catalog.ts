// (C) 2019 GoodData Corporation
import { GdcCatalog } from "../src";

export const catalogAttribute: GdcCatalog.ICatalogAttribute = {
    identifier: "attr",
    type: "attribute",
    title: "Attribute",
    summary: "",
    groups: [],
    production: true,
    links: {
        defaultDisplayForm: "/gdc/md/project/obj/1",
        self: "/gdc/md/project/obj/0",
    },
};

export const catalogMetric: GdcCatalog.ICatalogMetric = {
    identifier: "measure",
    type: "metric",
    title: "Measure",
    summary: "",
    groups: [],
    production: true,
    expression: "SELECT SUM([/gdc/md/project/obj/4])",
    format: "#,##",
    links: {
        self: "/gdc/md/project/obj/2",
    },
};

export const catalogFact: GdcCatalog.ICatalogFact = {
    identifier: "fact",
    type: "fact",
    title: "Fact",
    summary: "",
    groups: [],
    production: true,
    links: {
        self: "/gdc/md/project/obj/3",
    },
};
