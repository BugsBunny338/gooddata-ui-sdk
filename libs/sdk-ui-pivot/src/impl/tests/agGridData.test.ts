// (C) 2007-2020 GoodData Corporation

import { createIntlMock } from "@gooddata/sdk-ui";
import { createTableHeaders } from "../agGridHeaders";
import { getRow, getRowTotals } from "../agGridData";
import { DataViewFirstPage } from "@gooddata/sdk-backend-mockingbird";
import { ReferenceRecordings } from "@gooddata/reference-workspace";
import { recordedDataFacade } from "../../../__mocks__/recordings";

const intl = createIntlMock();

describe("getRowTotals", () => {
    it("should return total rows", () => {
        const fixture = recordedDataFacade(
            ReferenceRecordings.Scenarios.PivotTable.SingleMeasureAndMultipleGrandTotals,
            DataViewFirstPage,
        );

        const tableHeaders = createTableHeaders(fixture.dataView, { makeRowGroups: false });
        expect(
            getRowTotals(fixture, [...tableHeaders.rowFields, ...tableHeaders.colFields], intl),
        ).toMatchSnapshot();
    });
    it("should return null when no totals are defined", () => {
        const fixture = recordedDataFacade(
            ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
            DataViewFirstPage,
        );

        const tableHeaders = createTableHeaders(fixture.dataView, { makeRowGroups: false });
        expect(getRowTotals(fixture, [...tableHeaders.rowFields, ...tableHeaders.colFields], intl)).toBe(
            null,
        );
    });
});

describe("getRow", () => {
    it("should return a grid row", () => {
        const dv = recordedDataFacade(
            ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
            DataViewFirstPage,
        );

        const headerItems = dv.meta().allHeaders();
        const tableHeaders = createTableHeaders(dv.dataView, { makeRowGroups: false });

        expect(
            getRow(
                dv.twoDimData()[0],
                0,
                tableHeaders.colFields,
                tableHeaders.rowHeaders,
                headerItems[0],
                [],
                intl,
            ),
        ).toMatchSnapshot();
    });
    it("should return subtotal row", () => {
        const dv = recordedDataFacade(
            ReferenceRecordings.Scenarios.PivotTable.TwoMeasuresAndMultipleSubtotals,
            DataViewFirstPage,
        );

        const headerItems = dv.meta().allHeaders();
        const tableHeaders = createTableHeaders(dv.dataView, { makeRowGroups: false });

        expect(
            getRow(
                dv.twoDimData()[0],
                3,
                tableHeaders.colFields,
                tableHeaders.rowHeaders,
                headerItems[0],
                [null, null, null, "even"],
                intl,
            ),
        ).toMatchSnapshot();
    });
});
