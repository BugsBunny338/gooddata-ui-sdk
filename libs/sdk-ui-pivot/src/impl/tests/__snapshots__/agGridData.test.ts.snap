// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRow should return a grid row 1`] = `
Object {
  "a_1055": "CompuSci",
  "a_1086_460488-m_0": "5099309.54",
  "a_1086_460489-m_0": "22123590.1",
  "headerItemMap": Object {
    "a_1055": Object {
      "attributeHeaderItem": Object {
        "name": "CompuSci",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054/elements?id=165678",
      },
    },
  },
}
`;

exports[`getRow should return subtotal row 1`] = `
Object {
  "a_1055": "CompuSci",
  "a_1062_165680-a_1086_460488-m_0": "2174478.26",
  "a_1062_165680-a_1086_460488-m_1": "883793.08",
  "a_1062_165680-a_1086_460489-m_0": "5917286.73",
  "a_1062_165680-a_1086_460489-m_1": "2370306.08",
  "a_1062_165682-a_1086_460488-m_0": "1989393.54",
  "a_1062_165682-a_1086_460488-m_1": "817756.22",
  "a_1062_165682-a_1086_460489-m_0": "5501537.16",
  "a_1062_165682-a_1086_460489-m_1": "2039952.64",
  "a_1089": "Median",
  "headerItemMap": Object {
    "a_1055": Object {
      "attributeHeaderItem": Object {
        "name": "CompuSci",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054/elements?id=165678",
      },
    },
    "a_1089": Object {
      "totalHeaderItem": Object {
        "name": "med",
        "type": "med",
      },
    },
  },
  "subtotalStyle": null,
  "type": "rowSubtotal",
}
`;

exports[`getRowTotals should return total rows 1`] = `
Array [
  Object {
    "a_1055": "Sum",
    "a_1091_165679-a_1086_460488-m_0": "10542759.65",
    "a_1091_165679-a_1086_460489-m_0": "27767993.8",
    "a_1091_165872-a_1086_460488-m_0": "10404537.96",
    "a_1091_165872-a_1086_460489-m_0": "32066033.2",
    "a_1091_166425-a_1086_460488-m_0": "1445269.97",
    "a_1091_166425-a_1086_460489-m_0": "1161023.49",
    "a_1091_166442-a_1086_460488-m_0": "926586.28",
    "a_1091_166442-a_1086_460489-m_0": "2140879.84",
    "a_1091_166450-a_1086_460488-m_0": "667035.74",
    "a_1091_166450-a_1086_460489-m_0": "1194979.99",
    "a_1091_166498-a_1086_460488-m_0": "1813790.98",
    "a_1091_166498-a_1086_460489-m_0": "16633475.16",
    "a_1091_166647-a_1086_460488-m_0": "948454.04",
    "a_1091_166647-a_1086_460489-m_0": "3300573.84",
    "a_1091_168751-a_1086_460488-m_0": "1268661.8",
    "a_1091_168751-a_1086_460489-m_0": "4343400.8",
    "colSpan": Object {
      "count": 2,
      "headerKey": "a_1055",
    },
    "rowTotalActiveMeasures": Array [
      "m_0",
    ],
    "type": "rowTotal",
  },
  Object {
    "a_1055": "Max",
    "a_1091_165679-a_1086_460488-m_0": "2095977.22",
    "a_1091_165679-a_1086_460489-m_0": "4410258.72",
    "a_1091_165872-a_1086_460488-m_0": "1759859.66",
    "a_1091_165872-a_1086_460489-m_0": "6690298.08",
    "a_1091_166425-a_1086_460488-m_0": "751477.86",
    "a_1091_166425-a_1086_460489-m_0": "389676.45",
    "a_1091_166442-a_1086_460488-m_0": "627304.21",
    "a_1091_166442-a_1086_460489-m_0": "707188.64",
    "a_1091_166450-a_1086_460488-m_0": "398641.54",
    "a_1091_166450-a_1086_460489-m_0": "323372.36",
    "a_1091_166498-a_1086_460488-m_0": "770602.99",
    "a_1091_166498-a_1086_460489-m_0": "13387810.63",
    "a_1091_166647-a_1086_460488-m_0": "245096.79",
    "a_1091_166647-a_1086_460489-m_0": "1001196.79",
    "a_1091_168751-a_1086_460488-m_0": "324710.2",
    "a_1091_168751-a_1086_460489-m_0": "1116928.39",
    "colSpan": Object {
      "count": 2,
      "headerKey": "a_1055",
    },
    "rowTotalActiveMeasures": Array [
      "m_0",
    ],
    "type": "rowTotal",
  },
  Object {
    "a_1055": "Min",
    "a_1091_165679-a_1086_460488-m_0": "136710.01",
    "a_1091_165679-a_1086_460489-m_0": "669682.5",
    "a_1091_165872-a_1086_460488-m_0": "14990.06",
    "a_1091_165872-a_1086_460489-m_0": "485644.88",
    "a_1091_166425-a_1086_460488-m_0": "18000",
    "a_1091_166425-a_1086_460489-m_0": "3129.5",
    "a_1091_166442-a_1086_460488-m_0": "2735.24",
    "a_1091_166442-a_1086_460489-m_0": "3564.9",
    "a_1091_166450-a_1086_460488-m_0": "2339.1",
    "a_1091_166450-a_1086_460489-m_0": "16714.71",
    "a_1091_166498-a_1086_460488-m_0": "3601.6",
    "a_1091_166498-a_1086_460489-m_0": "14001.6",
    "a_1091_166647-a_1086_460488-m_0": "230.3",
    "a_1091_166647-a_1086_460489-m_0": "7827.29",
    "a_1091_168751-a_1086_460488-m_0": "1988",
    "a_1091_168751-a_1086_460489-m_0": "28203.56",
    "colSpan": Object {
      "count": 2,
      "headerKey": "a_1055",
    },
    "rowTotalActiveMeasures": Array [
      "m_0",
    ],
    "type": "rowTotal",
  },
  Object {
    "a_1055": "Rollup (Total)",
    "a_1091_165679-a_1086_460488-m_0": "10542759.65",
    "a_1091_165679-a_1086_460489-m_0": "27767993.8",
    "a_1091_165872-a_1086_460488-m_0": "10404537.96",
    "a_1091_165872-a_1086_460489-m_0": "32066033.2",
    "a_1091_166425-a_1086_460488-m_0": "1445269.97",
    "a_1091_166425-a_1086_460489-m_0": "1161023.49",
    "a_1091_166442-a_1086_460488-m_0": "926586.28",
    "a_1091_166442-a_1086_460489-m_0": "2140879.84",
    "a_1091_166450-a_1086_460488-m_0": "667035.74",
    "a_1091_166450-a_1086_460489-m_0": "1194979.99",
    "a_1091_166498-a_1086_460488-m_0": "1813790.98",
    "a_1091_166498-a_1086_460489-m_0": "16633475.16",
    "a_1091_166647-a_1086_460488-m_0": "948454.04",
    "a_1091_166647-a_1086_460489-m_0": "3300573.84",
    "a_1091_168751-a_1086_460488-m_0": "1268661.8",
    "a_1091_168751-a_1086_460489-m_0": "4343400.8",
    "colSpan": Object {
      "count": 2,
      "headerKey": "a_1055",
    },
    "rowTotalActiveMeasures": Array [
      "m_0",
    ],
    "type": "rowTotal",
  },
]
`;
